### ISSUE TITLE
Enhancements to linting rules

### ISSUE THREAD

**Issue by samdmarshall** [PyCQA/pylint#1005]
Hi! I installed pylint last night to add to one of my Python projects in an attempt to create a requirement of maintaining lint-free code when contributing. I was really impressed with the number of rules that pylint has built into it, but also found it very difficult to work out how to configure the settings. 
### Background

I have a lot of experience working in C-based languages and use a tool called [clang-format](http://clang.llvm.org/docs/ClangFormat.html) for maintaining style and lint-free code in my C/C++/Objective-C projects. This tool has a fair number of options and ways you can tweak the rules it uses to match however you want your project to be formatted. You can check out the various options available [here](http://clang.llvm.org/docs/ClangFormatStyleOptions.html).
### Enhancements

The major thing that stood out to me about clang-format vs pylint is that pylint has a lot of, what i might call "higher level" directives -- stuff that would be caught in C-style languages by the compiler rather than a linter. So while I find the options in pylint to be much more comprehensive in doing analysis on my code, I did find the ability to tweak the behavior of many of the options to be lacking.

For example, having an option to enable or disable the `trailing-whitespace` check for comment lines would be extremely useful to me. While I do care about extra trailing whitespace on code-lines, I don't care about trailing whitespace that is part of a comment line. Those results throw up false positives a report that can hide more serious issues with the way the code is written.

A number of the options, such as the formatting for method/function/class/variable/etc names, currently uses regex to set the expectations of behavior. I think that a number of built-in default patterns should be able to be used for these in addition to the custom regex rules. I would argue that the vast majority of projects that use pylint would use a handful of common naming conventions that could be standardized across projects. 

Another example of an issue i ran into was that my print statements are formatted as `print('some string')` in order to support compatibility between Python 2 and 3. However, pylint raises a warning on this for unnecessary parenthesis being used. I would rather not disable that linter option but it will give false-positives that actually break the code's intended compatibility with Python 3.

I don't have many other specific examples of enhancements that could be made because I'm still learning how to use pylint and the myriad of options that are available to me with it, but many of the settings could use some more fine-grain control with sets of defaults or optionally enforced parameters. One of the maintainers asked me on Twitter to open an issue about this so there could be some conversation about how these improvements might taken into consideration. I'm curious to know your thoughts about this.

**Comment by PCManticore** [comment]
Hi @samdmarshall !

Thank you for taking the time to write this feedback. It's immensely useful to see the way how our users interact with the tool and what it can be improved for easing their workflow.

I do agree that some of our rules are not that configurable as we intend them to be, especially coming from a tool that prides itself to be configurable to the bone. Usually they fall into the formatting category, bad-indentation, bad-continuation or other similar checks are falling short of being customized. 

Now to break this up, for the trailing whitespace problem, I created issue #1009.

For the formatting of the method/function/class names issue, what do you mean by built-in default patterns? An example would be useful.

For the issue regarding the unnecessary paranthesis for the print statement, there is a current solution. 
If you try to use ``from __future__ import print_function``, pylint will stop complaining on Python 2 that you are having extraneous paranthesis, since now they are part of the function call itself, while without the import, pylint thinks it is still a statement and that you are printing a tuple rather than calling the function. Having the import included is definitely a good practice, since without it, at least on Python 2, printing multiple objects will print a tuple of them, instead of printing the actual objects.

Let me know what other issues you might encounter. Thank you so much for reaching out. ;-)

**Comment by samdmarshall** [comment]
Sure! what I mean by built-in defaults is similar to some of the options in clang-format:

```
$ clang-format -style="{BreakBeforeBraces: Allman}" test.c
#include <stdio.h>

int main(int argc, char *argv[])
{
  int a = 1;
  if (a == 1)
  {
    printf("hello world!\n");
  }
  else
  {
    print("a = %i\n", a);
  }
  return 0;
}
```
vs.
```
$ clang-format -style="{BreakBeforeBraces: Stroustrup}" test.c
#include <stdio.h>

int main(int argc, char *argv[])
{
  int a = 1;
  if (a == 1) {
    printf("hello world!\n");
  }
  else {
    print("a = %i\n", a);
  }
  return 0;
}
```

The difference here being that the first case is using the Allman style of brace placement vs the second being the Stroustrup style. There are a couple of other styles for braces:
* Attach: Always attach braces to surrounding context.
* Linux: Like Attach, but break before braces on function, namespace and class definitions.
* Mozilla: Like Attach, but break before braces on enum, function, and record definitions.
* Stroustrup: Like Attach, but break before function definitions, catch, and else.
* Allman: Always break before braces.
* GNU: Always break before braces and add an extra level of indentation to braces of control statements, not to those of class, function or other definitions.
* WebKit: Like Attach, but break before functions.
* Custom: Configure each individual brace in `BraceWrapping`.

While clang-format has the options available for the user to completely customize how braces will be wrapped around every type of instance that would requires them -- there are a common set of styles that are used that can be defaulted to without having to manually tweak every value to achieve that. So in the case of the naming conventions, there are pretty common patterns such as snake case and camel case. Instead of having to input the parameters for all the naming options in regex, pre-made options could be setup to have regex patterns supplied by pylint instead of forcing that onto the user to create.

**Comment by moylop260** [comment]
I don't sure of understand the matter of use `styles` predefined.

Is a pre-configured template?

Similar to create example configuration files and use them with:
pylint --rcfile=example/gnu.cfg
pylint --rcfile=example/linux.cfg

Or a new short way
pylint --style=linux  # taking the example configuration file example/linux.cfg with custom disables and enables...

**Comment by samdmarshall** [comment]
Possibly, my initial thought was that the options in the configuration file would accept regex as they do right now, but also match against predefined strings. So, for example:

```
# Regular expression matching correct variable names
variable-rgx=[a-z_][a-z0-9_]{2,30}$

# Naming hint for variable names
variable-name-hint=[a-z_][a-z0-9_]{2,30}$
```
is the current way to define any variable using small_snake_case. What I am proposing is allowing pre-defined values to be passed which would be substituted for the correct pre-defined regexes. In this case something like:
```
# Regular expression matching correct variable names
variable-rgx=small_snake_case

# Naming hint for variable names
variable-name-hint=small_snake_case
```
where the `small_snake_case` value is a known pre-defined type within pylint so that it will correctly supply the appropriate regular expression for that naming convention.

**Comment by PCManticore** [comment]
This definitely makes sense, but I wouldn't conflate two meanings into variable-rgx, instead we should probably add a new option which can be used for defining a particular style. This is in line with what we did in the past whenever we had the choice of putting two semantics into the same option, e.g. we introduced separate options.

The problem would be with variable-name-hint, which by default is based on regexes, it should have been variable-name-hint-rgx instead, but we can probably use variable-name-hint-style or something of sorts for the the hint style.

I've created #1013 for handling this.
